racetime/views/race.py

```diff
@@ -160,6 +185,51 @@ def post(self, request, *args, **kwargs):
         return http.HttpResponse()
 
 
+@method_decorator(csrf_exempt, name='dispatch')
+class OAuthRaceChatDelete(ScopedProtectedResourceView, BotMixin, RaceChatMixin):
+    required_scopes = ['chat_delete']
+
+    def post(self, request, *args, **kwargs):
+        race = self.get_object()
+        message = self.get_message(race)
+
+        if message.is_system:
+            return http.JsonResponse({
+                'errors': ['System messages cannot be deleted.'],
+            }, status=422)
+
+        if not self.user.is_staff and race.chat_is_closed:
+            return http.JsonResponse({
+                'errors': [
+                    'This race chat is now closed. Please contact staff if '
+                    'you need to delete something.'
+                ],
+            }, status=422)
+
+        if not message.deleted:
+            message.deleted = True
+            message.deleted_by = self.user
+            message.deleted_at = timezone.now()
+            message.save()
+
+        channel_layer = get_channel_layer()
+        async_to_sync(channel_layer.group_send)(race.slug, {
+            'type': 'chat.delete',
+            'delete': {
+                'id': message.hashid,
+                'user': (
+                    message.user.api_dict_summary(race=race)
+                    if message.user else None
+                ),
+                'bot': message.bot.name if message.is_bot else None,
+                'is_bot': message.is_bot,
+                'deleted_by': self.user.api_dict_summary(race=race),
+            },
+        })
```

Comment:
There's obviously quite a bit of duplication here with `RaceChatDelete`, could you look into DRYing it up a little?